lib/puppet/application/ssl.rb

```diff
@@ -57,77 +87,111 @@ def main
+      name: host.name, server: Puppet[:ca_server], port: Puppet[:ca_port]
+    }
   rescue => e
-    puts "Failed to submit certificate request: #{e.message}"
-    exit(1)
+    raise Puppet::Error, _("Failed to submit certificate request: %{message}") % { message: e.message }
   end
 
   def download_cert(host)
     host.ensure_ca_certificate
 
-    puts "Downloading certificate '#{host.name}' from https://#{Puppet[:ca_server]}:#{Puppet[:ca_port]}"
-    if cert = host.download_host_certificate
-      puts "Downloaded certificate '#{host.name}' with fingerprint #{cert.fingerprint}"
-    else
-      puts "No certificate for '#{host.name}' on CA"
-    end
+    Puppet.info _("Downloading certificate '%{name}' from https://%{server}:%{port}") % {
+      name: host.name, server: Puppet[:ca_server], port: Puppet[:ca_port]
+    }
+    cert = host.download_host_certificate
+    return unless cert
+
+    Puppet.notice _("Downloaded certificate '%{name}' with fingerprint %{fingerprint}") % {
+      name: host.name, fingerprint: cert.fingerprint
+    }
+    cert
   rescue => e
-    puts "Failed to download certificate: #{e.message}"
-    exit(1)
+    raise Puppet::Error, _("Failed to download certificate: %{message}") % { message: e.message }
   end
 
   def verify(host)
     host.ensure_ca_certificate
 
     key = host.key
-    unless key
-      puts "The host's private key is missing"
-      exit(1)
-    end
+    raise _("The host's private key is missing") unless key
 
     cert = host.check_for_certificate_on_disk(host.name)
-    unless cert
-      puts "The host's certificate is missing"
-      exit(1)
-    end
+    raise _("The host's certificate is missing") unless cert
 
     if cert.content.public_key.to_pem != key.content.public_key.to_pem
-      puts "The host's key does not match the certificate"
-      exit(1)
+      raise _("The host's key does not match the certificate")
     end
 
     store = host.ssl_store
     unless store.verify(cert.content)
-      puts "Failed to verify certificate '#{host.name}': #{store.error_string} (#{store.error})"
-      exit(1)
+      raise _("Failed to verify certificate '%{name}': %{message} (%{error})") % {
+        name: host.name, message: store.error_string, error: store.error
+      }
     end
 
-    puts "Verified certificate '#{host.name}'"
+    Puppet.notice _("Verified certificate '%{name}'") % {
+      name: host.name
+    }
     # store.chain.reverse.each_with_index do |issuer, i|
     #   indent = "  " * (i+1)
-    #   puts "#{indent}#{issuer.subject.to_s}"
+    #   Puppet.notice "#{indent}#{issuer.subject.to_s}"
     # end
-    exit(0)
   rescue => e
-    puts "Verify failed: #{e.message}"
-    exit(1)
+    raise Puppet::Error, _("Verify failed: %{message}") % { message: e.message }
+  end
+
+  def clean(host)
+    if Puppet[:certname] == Puppet[:ca_server]
+      # make sure cert has been removed from the CA
+      cert = host.download_certificate_from_ca(Puppet[:certname])
```

Comment:
This will either return the cert, nil (on 404) or raise if there was another issue. Do we need to handle any of those?
Most of them seem benign to let raise, though I don't know if we talked about this case (your CA's offline)?
```
puppet justin (mergejosh $) ‖ be puppet ssl clean
Error: Could not run: Failed to open TCP connection to localhost:8140 (Connection refused - connect(2) for "localhost" port 8140)
```

Reply:
yeah let me rescue that and print a better message. As it is, it's not clear why "ssl clean" would be making a network request.

Reply:
@justinstoller updated